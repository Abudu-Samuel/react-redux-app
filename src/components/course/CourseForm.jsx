import React from 'react';
import PropTypes from 'prop-types';
import TextInput from '../common/TextInput';
import SelectInput from '../common/SelectInput';

const CourseForm = ({
  course, allAuthors, onSave, onChange, errors, isSaving, courseToModify
}) => (
  <form>
    <TextInput
        name="title"
        label="Title"
        defaultValue={courseToModify.title}
        onChange={onChange}
        placeholder="Title"
        error={errors.title}
      />
    <SelectInput
    name="authorId"
    label="Author"
    defaultValue={courseToModify.authorId}
    defaultOption="Select Author"
    options={allAuthors}
    onChange={onChange}
    error={errors.authorId}
    />

    <TextInput
      name="category"
      label="Category"
      placeholder="Category"
      defaultValue={courseToModify.category}
      onChange={onChange}
      error={errors.category}
    />

    <TextInput
      name="length"
      label="Length"
      placeholder="Length"
      onChange={onChange}
      defaultValue={courseToModify.length}
      error={errors.length}
    />

    <input
      name="submit"
      disabled={isSaving}
      defaultValue={isSaving ? 'saving' : 'Save'}
      className="btn btn-sm btn-primary"
      onClick={onSave}
    />
  </form>
);

CourseForm.propTypes = {
  course: PropTypes.object.isRequired,
  allAuthors: PropTypes.array.isRequired,
  onSave: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  isSaving: PropTypes.bool.isRequired,
  errors: PropTypes.object.isRequired
};

export default CourseForm;