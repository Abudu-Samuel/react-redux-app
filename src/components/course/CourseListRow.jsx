import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

const CourseListRow = ({ course, handleDelete }) => (
  <tr>
    <td><a href={course.watchHref}>Watch</a></td>
    <td><Link to={`/course/${course.id}`}>{course.title}</Link></td>
    <td>{course.authorId}</td>
    <td>{course.category}</td>
    <td>{course.length}</td>
    <td><button onClick={() => handleDelete(course)} className="btn-danger" type="button">Delete</button></td>
  </tr>
);

CourseListRow.propTypes = {
  course: PropTypes.object.isRequired
};

export default CourseListRow;
