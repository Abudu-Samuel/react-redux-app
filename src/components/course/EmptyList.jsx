import React from 'react';

const EmptyList = () => (
  <div>
    <p className="text-center"><i className="fa fa-exclamation-circle " /></p>
    <p className="text-center">There are no courses</p>
  </div>
);

export default EmptyList;