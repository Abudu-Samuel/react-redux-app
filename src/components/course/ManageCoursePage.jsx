/* eslint-disable react/destructuring-assignment */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import * as actions from '../../actions/courseActions';
import CourseForm from './CourseForm';
import validateInput from '../../helper/validation';

class ManageCoursePage extends Component {
  state = {
    errors: {},
    title: '',
    category: '',
    authorId: '',
    length: ''
  }

  componentWillReceiveProps = (nextProps) => {
    if (this.props.course.id !== nextProps.course.id) {
      this.setState({
        course: nextProps.course
      });
    }
  }

  updateCourseState = (event) => {
    const { name, value } = event.target;
    // eslint-disable-next-line react/no-access-state-in-setstate
    const { course } = this.props;
    if (course.authorId !== '') {
      this.setState({
        ...course
      });
    }
    this.setState({ [name]: value });
  }

  saveCourse = (event) => {
    event.preventDefault();
    const { course } = this.props;
    if (this.state.authorId === '' && course.authorId !== '') {
      this.setState({
        ...course
      }, () => this.validateAndDispatchAction());
      return;
    }
    this.validateAndDispatchAction();
  }

  validateAndDispatchAction = () => {
    const validateErrors = validateInput({
      ...this.state,
      title: this.state.title.split(' ').join(''),
      authorId: this.state.authorId.split(' ').join('')
    });
    if (Object.keys(validateErrors).length > 0) {
      this.setState({ errors: validateErrors });
      return;
    }
    this.setState({ errors: {} });
    this.props.actions.saveCourse(this.state)
      .then(() => {
        this.props.history.push('/courses');
      });
  }

  render() {
    const { errors } = this.state;
    const { authors, isSaving, course } = this.props;
    return (
      <div>
        <h1>Manage course</h1>
        <CourseForm
          allAuthors={authors}
          onChange={this.updateCourseState}
          onSave={this.saveCourse}
          course={this.state}
          courseToModify={course}
          isSaving={isSaving}
          errors={errors}
        />
      </div>
    );
  }
}

ManageCoursePage.propTypes = {
  course: PropTypes.object.isRequired,
  authors: PropTypes.array.isRequired,
  actions: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  isSaving: PropTypes.bool.isRequired
};

const getCourseById = (courses, id) => {
  const course = courses.allCourses.courses.filter(course => course.id === id);
  if (course) return course[0];
  return null;
};

const mapStateToProps = (state, ownProps) => {
  const courseId = ownProps.match.params.id;
  let course = {
    id: '', watchHref: '', title: '', authorId: '', length: '', category: ''
  };

  if (state.courses.allCourses.courses) {
    if (courseId && state.courses.allCourses.courses.length > 0) {
      course = getCourseById(state.courses, courseId);
    }
  }

  const authorsFormattedForDropdown = state.authors.allAuthors.map(author => ({
    value: author.id,
    text: `${author.firstName} ${author.lastName}`
  }));

  return {
    course,
    authors: authorsFormattedForDropdown,
    isSaving: state.courses.isSaving
  };
};

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(actions, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(ManageCoursePage);