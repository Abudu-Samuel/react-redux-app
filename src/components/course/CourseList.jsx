import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';
import CourseListRow from './CourseListRow';
import EmptyList from './EmptyList';

const CourseList = ({
  courses, handleDelete, currentPage, loadCourses, totalCourses, handlePageChange
}) => (
  <Fragment>
    {
      totalCourses > 0
        ? (
          <div>
            <table className="table">
              <thead>
                <tr>
                  <th>&nbsp;</th>
                  <th>Title</th>
                  <th>Author</th>
                  <th>Category</th>
                  <th>Length</th>
                  <th>action</th>
                </tr>
              </thead>
              <tbody>
                {
            courses.map(course => (
              <CourseListRow key={course.id} course={course} handleDelete={handleDelete} />
            ))
          }
              </tbody>
            </table>
            <ReactPaginate
        pageCount={Math.ceil(totalCourses / 2)}
        pageRangeDisplayed={2}
        onPageChange={event => handlePageChange(event)}
        marginPagesDisplayed={1}
        initialPage={currentPage - 1}
        disableInitialCallback
      />
            {' '}

          </div>
        )
        : <EmptyList />
    }


  </Fragment>
);

CourseList.propTypes = {
  courses: PropTypes.array.isRequired
};

export default CourseList;