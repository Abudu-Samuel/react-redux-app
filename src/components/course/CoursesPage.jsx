import React, { Component } from 'react';
import { connect } from 'react-redux';
import Loader from 'react-md-spinner';
import CourseList from './CourseList';
import { deleteCourse, loadCourses } from '../../actions/courseActions';

class CoursesPage extends Component {
  componentDidMount() {
    this.props.loadCourses();
  }

  componentDidUpdate() {
    const { totalCourses, currentPage, courses, loadCourses } = this.props;
    if (courses.length === 0 && currentPage > 1) {
      loadCourses(currentPage - 1);
    }
    if (courses.length === 0 && currentPage === 1) {
      loadCourses(currentPage);
    }
  }

  handlePageChange = ({ selected }) => {
    const { loadCourses } = this.props;
    loadCourses(selected + 1);
  }

  redirectToAddCoursePage = () => {
    const { history } = this.props;
    history.push('/course');
  }

  render() {
    const {
      courses, isFetching, currentPage, totalCourses
    } = this.props;

    return (
      <div>
        <h1>Courses</h1>
        <input
          type="submit"
          value="Add course"
          className="btn btn-primary"
          onClick={this.redirectToAddCoursePage}
        />
        { ' ' }
        <button className="btn btn-primary" type="button">{totalCourses}</button>
        {
          isFetching || !courses ? <Loader />
            : (
              <CourseList
                totalCourses={totalCourses}
                currentPage={currentPage}
                courses={courses}
                handleDelete={this.props.deleteCourse}
                handlePageChange={this.handlePageChange}
              />
            )
        }
      </div>
    );
  }
}

const mapStateToProps = ({
  courses: { allCourses: { courses, currentPage, totalCourses, loadCourses }, isFetching }
}) => ({
  courses,
  isFetching,
  currentPage,
  totalCourses
});


export default connect(mapStateToProps, { deleteCourse, loadCourses })(CoursesPage);