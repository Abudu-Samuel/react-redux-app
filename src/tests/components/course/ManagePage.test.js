/* eslint-disable */
import { ManageCoursePage } from '../../../components/course/ManageCoursePage';

const props = {
  authors: [{}],
  isSaving: false,
  course: [{}],
  actions: {
    saveCourse: jest.fn(() => Promise.resolve()),
  }
};

const state = {
  errors: {},
  title: '',
  category: '',
  authorId: '',
  length: '',
  noEdit: false
}


const event = {
  target: {
    name: 'title',
    value: 'some title'
  },
  preventDefault: jest.fn(),
};


const wrapper = shallow(<ManageCoursePage {...props} />);

describe('renders properly', () => {
  it('render ManageCoursePage component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('trigger handleEvent method', () => {
    const instance = wrapper.instance();
    instance.updateCourseState({ target: { name: 'title', value: 'INTRO' } });
    expect(instance.state.title).toEqual('INTRO');
  });

  it('trigger saveCourse', () => {
    wrapper.setState({
      authorId: 'cory-house',
      title: 'Advanced redux',
      category: 'new one',
      length: '23:05'
    })
    const instance = wrapper.instance();
    instance.saveCourse(event);
    expect(props.actions.saveCourse).toBeCalled();
  });
});