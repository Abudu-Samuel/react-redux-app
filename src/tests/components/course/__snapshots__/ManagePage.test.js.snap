// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly render ManageCoursePage component 1`] = `
<div>
  <h1>
    Manage course
  </h1>
  <CourseForm
    allAuthors={
      Array [
        Object {},
      ]
    }
    course={
      Object {
        "authorId": "",
        "category": "",
        "errors": Object {},
        "length": "",
        "noEdit": false,
        "title": "",
      }
    }
    courseToModify={
      Array [
        Object {},
      ]
    }
    errors={Object {}}
    isSaving={false}
    onChange={[Function]}
    onSave={[Function]}
  />
</div>
`;
